### PR TITLE
Clean ra loop

### PR DESCRIPTION
**Pull request by tomasbrod**
* remove ResearchAverageMagnitude
* remove mvCPIDBlockHashes
* add vRewardBlocs set of blockindex into StructCPID
* clean up declarations
* use const reference for GetInitializedStructCPID2
* remove AddCPIDBlockHash
* move and deduplicate ResearchAverageMagnitude calc to AddRARewardBlock
* move GetLifetimeCPID loop into separate RescanLifetimeCPID (called)
* change some warnings into hard asserts
* change some RPC outputs
* warn and reload index cpids where cpid is 0000
* do not add blocks to RA where RA is not enabled
* include chain head (best) to RA scan on load

The PR is made on top of master, but I do not suggest merging it into master directly.

### CODE REVIEW

File: src/main.cpp [hunk 1]
```diff
@@ -5708,89 +5677,90 @@ bool GetEarliestStakeTime(std::string grcaddress, std::string cpid)
...
+        stCPID . vRewardBlocs . emplace(pindex);
+        // Store the updated struct. (implicit)
+    }
 }
 
-void RemoveCPIDBlockHash(const std::string& cpid, const uint256& blockhash)
+void RemoveCPIDBlockHash(const std::string& cpid, const CBlockIndex* pindex)
 {
-   mvCPIDBlockHashes[cpid].erase(blockhash);
+    //TODO: is copying involved here?
```
Comment (denravonska) [inline src/main.cpp:5717]: Copy is indeed involved here. We could perhaps return a reference so we don't constantly have to Get->Update->Set when handling StructCPID objects.

Edit: Oh, you already fixed that.

**Comment by tomasbrod** [comment]
From ravon about the index cpid inconsistency:
```
diff --git a/src/txdb-leveldb.cpp b/src/txdb-leveldb.cpp
index bd001fd07..93dd1e763 100755
--- a/src/txdb-leveldb.cpp
+++ b/src/txdb-leveldb.cpp
@@ -25,6 +25,7 @@ using namespace boost;
 
 leveldb::DB *txdb; // global pointer for LevelDB object instance
 void AddCPIDBlockHash(const std::string& cpid, const uint256& blockhash);
+MiningCPID GetBoincBlockByIndex(CBlockIndex* pblockindex);
 
 static leveldb::Options GetOptions() {
     leveldb::Options options;
@@ -638,6 +639,20 @@ bool CTxDB::LoadBlockIndex()
                 }
             }            
 
+            // Fix for an old bug which stored CPIDs as 0x00.
+            if(pindex->IsUserCPID() && pindex->cpid == uint128())
+            {
+                const MiningCPID& cpid = GetBoincBlockByIndex(pindex);
+                if(cpid.initialized)
+                {
+                    pindex->SetCPID(cpid.cpid);
+                    LogPrintf("Repaired block %s with CPID %s", pindex->GetBlockHash().GetHex(), cpid.cpid);
+
+                    if(!WriteBlockIndex(CDiskBlockIndex(pindex)))
+                        LogPrintf("Error writing block to disk");
+                }
+            }
+
             if (pindex->nResearchSubsidy > 0 && pindex->IsUserCPID())
             {
                 const std::string& scpid = pindex->GetCPID();
```

**Comment by jamescowens** [comment]
Looks good to me. utACK.

File: src/main.cpp (hunk 1 above)
Comment (jamescowens) [reply src/main.cpp:5717]: He is fast! Makes me feel like a slug! :)

**Pull request merged**